File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/32 bits/xorshift32.pas
{ PRNG .......: XorShift (32 bits)
  Ano ........: 2003
  Autor ......: George Marsaglia 
}

program XorShift32PRNG;

uses
  SysUtils;

type
  xorShift32State = record
    seed: UInt32;
  end;

function xorShift32(var state: xorShift32State): UInt32;
var
  random: UInt32;
begin
  random := state.seed;
  random := random xor (random shl 13);
  random := random xor (random shr 17);
  random := random xor (random shl 5);
  state.seed := random;
  xorShift32 := state.seed;
end;

var
  Xn1, Xn2: xorShift32State;
  i: Integer;
  
begin

  Xn1.seed := 1234;
  Xn2.seed := UInt32(GetTickCount);

  Writeln('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  Writeln('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  Writeln('PADRAO: XorShift (32 bits)');
  Writeln('=============================================');
  Writeln;
  Writeln('      Semente: 1234       Semente: TIME');
  Writeln('      -------------       -------------');

  for i := 1 to 10 do
    Writeln(Format('%19u%20u', [xorShift32(Xn1), xorShift32(Xn2)]));

end.

File: Numeros Pseudo-Aleatorios/XorShift 32, 64 e 128 bits/32 bits/xorshift32.go
/*
 PRNG .......: XorShift (64 bits)
 Ano ........: 2003
 Autor ......: George Marsaglia
*/

package main

import (
    "fmt"
    "time"
)

type xorShift32State struct {
    seed uint32
}

func xorshift32(state *xorShift32State) uint32 {
    random := state.seed
    random ^= random << 13
    random ^= random >> 17
    random ^= random << 5
    state.seed = random
    return state.seed
}

func main() {
	
    var Xn1 xorShift32State
    var Xn2 xorShift32State

    Xn1.seed = 1234
    Xn2.seed = uint32(time.Now().Unix())

    fmt.Println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
    fmt.Println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
    fmt.Println("PADRAO: XorShift (32 bits)")
    fmt.Println("=============================================")
    fmt.Println()
    fmt.Println("      Semente: 1234       Semente: TIME")
    fmt.Println("      -------------       -------------")

    for i := 1; i <= 10; i++ {
        fmt.Printf("%19d%20d\n", xorshift32(&Xn1), xorshift32(&Xn2))
    }
}
